src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs

```diff
@@ -1748,102 +1817,95 @@ private void PassivateCompleted(EntityId entityId)
             }
         }
 
-        private void DeliverMessage(object msg, IActorRef snd)
+        private void DeliverMessage(string entityId, object msg, IActorRef snd)
         {
-            var t = _extractEntityId(msg);
-            var entityId = t.Value.Item1;
-            var payload = t.Value.Item2;
+            var payload = _extractor.EntityMessage(msg)!; // payload can't be null unless dev really screwed up
 
-            if (string.IsNullOrEmpty(entityId))
-            {
-                Log.Warning("{0}: Id must not be empty, dropping message [{1}]", _typeName, msg.GetType().Name);
-                Context.System.DeadLetters.Tell(new Dropped(msg, "No recipient entity id", snd, Self));
-            }
-            else
+            if (payload is ShardRegion.StartEntity start)
             {
-                switch (payload)
+                // Handling StartEntity both here and in the receives allows for sending it both as is and in an envelope
+                // to be extracted by the entity id messageExtractor.
+
+                // we can only start a new entity if we are not currently waiting for another write
+                if (_entities.PendingRememberedEntitiesExist)
+                {
+                    if (_verboseDebug)
+                        Log.Debug("{0}: StartEntity({1}) from [{2}], adding to batch", _typeName,
+                            start.EntityId, snd);
+                    _entities.RememberingStart(entityId, ackTo: snd);
+                }
+                else
                 {
-                    case ShardRegion.StartEntity start:
-                        // Handling StartEntity both here and in the receives allows for sending it both as is and in an envelope
-                        // to be extracted by the entity id extractor.
+                    if (_verboseDebug)
+                        Log.Debug("{0}: StartEntity({1}) from [{2}], starting", _typeName, start.EntityId, snd);
+                    StartEntity(start.EntityId, snd);
+                }
 
-                        // we can only start a new entity if we are not currently waiting for another write
+                return;
+            }
+
+            switch (_entities.EntityState(entityId))
```

Comment:
I think this could probably be perf-optimized, but that's a separate and distinct issue from what this PR is trying to achieve. cc @to11mtm